File: CPP/ByRound/1846/1846_G_Rudolf_and_CodeVid_23.cpp
// Problem: CF 1846 G - Rudolf and CodeVid-23
// https://codeforces.com/contest/1846/problem/G

/*
Algorithm: Shortest Path with Dijkstra
Approach: 
- Each symptom state is a node in a graph
- Each medicine is an edge with cost equal to the number of days it takes
- State transition: remove symptoms (a), then add side effects (b)
- Use Dijkstra's algorithm to find minimum days to reach state 0 (no symptoms)

Time Complexity: O(2^n * m * log(2^n)) where n <= 10, so 2^n <= 1024
Space Complexity: O(2^n) for dp and vis arrays, and O(m) for medicine data
*/

#include <algorithm>
#include <array>
#include <functional>
#include <iostream>
#include <queue>
#include <random>
#include <utility>
#include <vector>

#define fi first
#define se second
#define pb push_back
#define all(x) (x).begin(), (x).end()
#define SZ(x) ((int)x.size())
#define rep(i, a, n) for (int i = a; i < n; i++)
#define per(i, a, n) for (int i = n - 1; i >= a; i--)
#define IOS                                                                    \
  ios::sync_with_stdio(0);                                                     \
  cin.tie(0);                                                                  \
  cout.tie(0);
using namespace std;
using ll = long long;
using PII = pair<int, int>;
mt19937 mrand(random_device{}());
const int N = 1000010, M = 1000, mod = 1e9 + 7, inf = 1 << 30;
int n, m;
void solve() {
  cin >> n >> m;
  vector<int> dp(1 << n, inf), vis(1 << n); // dp[x] = min days to reach state x, vis[x] = visited flag
  vector<array<int, 3>> e(m); // e[i][0] = days, e[i][1] = symptoms removed, e[i][2] = side effects
  int s = 0; // initial state (bitmask of symptoms)
  rep(i, 0, n) {
    char c;
    cin >> c;
    s = s * 2 + c - '0'; // convert string to bitmask
  }
  rep(i, 0, m) {
    cin >> e[i][0]; // days needed
    int s = 0;
    rep(j, 0, n) {
      char c;
      cin >> c;
      s = s * 2 + c - '0'; // symptoms removed
    }
    e[i][1] = s;
    s = 0;
    rep(j, 0, n) {
      char c;
      cin >> c;
      s = s * 2 + c - '0'; // side effects
    }
    e[i][2] = s;
  }
  dp[s] = 0; // initial state requires 0 days
  priority_queue<PII, vector<PII>, greater<PII>> q; // priority queue for Dijkstra: {days, state}
  q.push({0, s});
  auto print = [&](int u) {
    vector<int> a(n);
    for (int i = n - 1; i >= 0; i--)
      a[i] = u >> i & 1;
    rep(i, 0, n) cout << a[i];
    cout << ' ' << dp[u] << endl;
  };
  while (q.size()) {
    auto [w, u] = q.top();
    q.pop();
    if (vis[u])
      continue; // Skip if already processed
    vis[u] = 1;
    for (auto [c, a, b] : e) { // For each medicine
      int v = (u & ~a) | b; // Apply medicine: remove symptoms (a), then add side effects (b)
      if (dp[v] > w + c) { // If we found a better path
        dp[v] = c + w;
        q.push({dp[v], v}); // Add to queue for processing
      }
    }
  }
  if (dp[0] >= inf) // If state 0 (no symptoms) is unreachable
    dp[0] = -1;
  cout << dp[0] << '\n';
}
int main() {
  IOS int t = 1;
  cin >> t;
  while (t--)
    solve();
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/G/1846_G_Rudolf_and_CodeVid_23.py
# Problem: CF 1846 G - Rudolf and CodeVid-23
# https://codeforces.com/contest/1846/problem/G

"""
Code Purpose:
This program finds the minimum number of days to eliminate all symptoms of the "CodeVid-23" virus by applying medicines optimally.
Each medicine has a duration, symptoms it removes, and side effects (symptoms it causes).
The problem is modeled as a shortest path in a graph where nodes represent symptom states (bitmask) and edges represent applying a medicine.
Dijkstra's algorithm is used to find the minimum number of days to reach the state with no symptoms.

Algorithms/Techniques:
- Bitmasking to represent symptom states
- Dijkstra's shortest path algorithm with a priority queue
- State transition using bitwise operations for applying medicines

Time Complexity: O(2^n * m * log(2^n)) where n is the number of symptoms and m is the number of medicines
Space Complexity: O(2^n) for the distance array and priority queue
"""

import sys

input = sys.stdin.readline
from heapq import heappop, heappush


def LII():
    return list(map(int, input().strip().split()))


inf = 10**7


def main():
    n, m = LII()
    # Convert initial symptoms from binary string to integer bitmask
    t = int(input(), 2)
    # Read medicines: [duration, removes_mask, side_effects_mask]
    A = [[int(input()), int(input(), 2), int(input(), 2)] for _ in range(m)]
    # Distance array to track minimum days to reach each symptom state
    d = [inf] * (1 << n)
    d[0] = 0  # No symptoms, 0 days needed
    pq = [(0, 0)]  # Priority queue: (days, symptom_mask)
    while pq:
        w, u = heappop(pq)
        # If all symptoms are gone, return the number of days
        if u & t == t:
            return w
        # Skip if we already found a better path to this state
        if w > d[u]:
            continue
        # Try each medicine
        for w2, a, b in A:
            # Compute new symptom state after applying medicine
            v = u | a  # Apply removals
            # Check if all side effects are present (can't apply if they are)
            if u & b == b and w + w2 < d[v]:
                d[v] = w + w2
                heappush(pq, (w + w2, v))
    return -1


for _ in range(int(input())):
    print(main())


# https://github.com/VaHiX/CodeForces/